### PR TITLE
Add custom logits processor API

### PR DESCRIPTION
**Pull request by EricLBuehler**
New sampling order:

1) Apply penalties from sampling_params
2) ⭐ Apply these custom logits processors sequentially
3) Apply temperature and softmax
4) Sample the next token (topk, topp, minp, etc)

@Dan-wanna-M would this API work? You can find an example of using it [here](https://github.com/EricLBuehler/mistral.rs/blob/9078786881b229f1d4f740b9cc570ee46be4e860/mistralrs/examples/custom_logits_processor/main.rs#L81-L83). Basically, you provide a Vec of `Arc<dyn Fn(&Tensor, &[u32]) -> Result<Tensor> + Send + Sync>`, which are applied. 

Please let me know if this works for you or what changes would be necessary, and I can merge this PR. [kbnf](https://github.com/Dan-wanna-M/kbnf) looks super cool!

### CODE REVIEW

**Comment by Dan-wanna-M** [comment]
@EricLBuehler One thing I am pondering is that `kbnf` is stateful, which means to use `Arc<dyn Fn()>` I need to use `Mutex<T>`. This requirement by itself is fine, but I am wondering its implications. For example, if logits processors will be directly shared across threads and executed in parallel, then some very bad stuff probably will happen.

**Comment by EricLBuehler** [comment]
> @EricLBuehler One thing I am pondering is that kbnf is stateful, which means to use Arc<dyn Fn()> I need to use Mutex<T>. This requirement by itself is fine, but I am wondering its implications. For example, if logits processors will be directly shared across threads and executed in parallel, then some very bad stuff probably will happen.

@Dan-wanna-M yes, we parallelize the sampling process. Perhaps we could change the API to accept some generic type which has some method to do the logits processing? Then, you could contain the state in that type in `Mutex`.

**Comment by EricLBuehler** [comment]
@Dan-wanna-M I just updated it so now the API is:
```rust
pub trait CustomLogitsProcessor: Send + Sync {
    /// Logits and sequence context (prompt and generated tokens), returning modified tokens.
    fn apply(&self, logits: &Tensor, context: &[u32]) -> Result<Tensor>;
}
```
In the example, we have a custom struct:
```rust
struct ThresholdLogitsProcessor {
    threshold: f64,
}

impl CustomLogitsProcessor for ThresholdLogitsProcessor {
    fn apply(&self, logits: &Tensor, _context: &[u32]) -> Result<Tensor> {
        // Mask is 1 for true, 0 for false.
        let mask = logits.ge(self.threshold)?;
        logits.broadcast_mul(&mask.to_dtype(logits.dtype())?)
    }
}
```

Would this work if you could put shared state into the custom struct w/ a mutex?

**Comment by Dan-wanna-M** [comment]
> @Dan-wanna-M I just updated it so now the API is:
> 
> ```rust
> pub trait CustomLogitsProcessor: Send + Sync {
>     /// Logits and sequence context (prompt and generated tokens), returning modified tokens.
>     fn apply(&self, logits: &Tensor, context: &[u32]) -> Result<Tensor>;
> }
> ```
> 
> In the example, we have a custom struct:
> 
> ```rust
> struct ThresholdLogitsProcessor {
>     threshold: f64,
> }
> 
> impl CustomLogitsProcessor for ThresholdLogitsProcessor {
>     fn apply(&self, logits: &Tensor, _context: &[u32]) -> Result<Tensor> {
>         // Mask is 1 for true, 0 for false.
>         let mask = logits.ge(self.threshold)?;
>         logits.broadcast_mul(&mask.to_dtype(logits.dtype())?)
>     }
> }
> ```
> 
> Would this work if you could put shared state into the custom struct w/ a mutex?

I spent some time rethinking it and I think as long as one generation only uses one logits processor then it should work(So a batch of N generation will need N separate logits processor)

**Comment by EricLBuehler** [comment]
> > @Dan-wanna-M I just updated it so now the API is:
> > ```rust
> > pub trait CustomLogitsProcessor: Send + Sync {
> >     /// Logits and sequence context (prompt and generated tokens), returning modified tokens.
> >     fn apply(&self, logits: &Tensor, context: &[u32]) -> Result<Tensor>;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In the example, we have a custom struct:
> > ```rust
> > struct ThresholdLogitsProcessor {
> >     threshold: f64,
> > }
> > 
> > impl CustomLogitsProcessor for ThresholdLogitsProcessor {
> >     fn apply(&self, logits: &Tensor, _context: &[u32]) -> Result<Tensor> {
> >         // Mask is 1 for true, 0 for false.
> >         let mask = logits.ge(self.threshold)?;
> >         logits.broadcast_mul(&mask.to_dtype(logits.dtype())?)
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Would this work if you could put shared state into the custom struct w/ a mutex?
> 
> I spent some time rethinking it and I think as long as one generation only uses one logits processor then it should work(So a batch of N generation will need N separate logits processor)

Sounds good, and this API implements `CustomLogitsProcessor` for the corresponding closure type! I'll merge this if there are no other problems.

**Comment by Dan-wanna-M** [comment]
> > > @Dan-wanna-M I just updated it so now the API is:
> > > ```rust
> > > pub trait CustomLogitsProcessor: Send + Sync {
> > >     /// Logits and sequence context (prompt and generated tokens), returning modified tokens.
> > >     fn apply(&self, logits: &Tensor, context: &[u32]) -> Result<Tensor>;
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > In the example, we have a custom struct:
> > > ```rust
> > > struct ThresholdLogitsProcessor {
> > >     threshold: f64,
> > > }
> > > 
> > > impl CustomLogitsProcessor for ThresholdLogitsProcessor {
> > >     fn apply(&self, logits: &Tensor, _context: &[u32]) -> Result<Tensor> {
> > >         // Mask is 1 for true, 0 for false.
> > >         let mask = logits.ge(self.threshold)?;
> > >         logits.broadcast_mul(&mask.to_dtype(logits.dtype())?)
> > >     }
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Would this work if you could put shared state into the custom struct w/ a mutex?
> > 
> > 
> > I spent some time rethinking it and I think as long as one generation only uses one logits processor then it should work(So a batch of N generation will need N separate logits processor)
> 
> Sounds good, and this API implements `CustomLogitsProcessor` for the corresponding closure type! I'll merge this if there are no other problems.

Yes it should be mergeable now

**Comment by EricLBuehler** [comment]
Sounds good @Dan-wanna-M! Just merged it.
